tee-worker/omni-executor/aa-contracts-client/src/entry_point_client.rs

```diff
@@ -247,6 +335,112 @@ impl<P: RpcProvider<Transaction = TransactionRequest, Addr = Address>> EntryPoin
 			.map_err(|_| error!("Could not send tx"))
 	}
 
+	/// Submit UserOperations with automatic retry and gas escalation on failure
+	pub async fn handle_ops_with_retry(
+		&self,
+		user_ops: &[PackedUserOperation],
+		beneficiary: Address,
+	) -> Result<String, ()> {
+		let mut attempt = 0;
+		let mut gas_buffer_adjustment = 0u64;
+
+		loop {
+			// Calculate gas fees with increasing buffer on retries
+			let (base_max_fee, base_priority_fee) =
+				self.calculate_gas_fees_with_buffer(gas_buffer_adjustment).await?;
+
+			// Build transaction with adjusted gas
+			let ops = user_ops.to_vec();
+			let call_data = handleOpsCall { ops, beneficiary }.abi_encode();
+			let mut tx = build_call_transaction(self.entry_point_address, call_data);
+
+			// Apply gas fees to transaction
+			tx.max_fee_per_gas = Some(base_max_fee.to::<u128>());
+			tx.max_priority_fee_per_gas = Some(base_priority_fee.to::<u128>());
+
+			// Attempt to send transaction
+			match self.rpc_client.send_transaction(tx).await {
+				Ok(tx_hash) => {
+					if attempt > 0 {
+						info!(
+							"Transaction succeeded after {} retries with gas buffer {}%",
+							attempt, gas_buffer_adjustment
+						);
+					}
+					return Ok(tx_hash);
+				},
+				Err(_) => {
+					// Since RPC provider returns unit error, we treat all errors as potentially gas-related
+					// TODO: improve error handling, classify retryable vs non-retryable errors
+
+					attempt += 1;
+					if attempt >= self.retry_config.max_attempts {
+						error!("Transaction failed after {} attempts", attempt);
+						return Err(());
+					}
+
+					// Calculate backoff delay
+					let delay = self.calculate_backoff_delay(attempt);
+
+					// Increase gas buffer for next attempt
+					gas_buffer_adjustment += self.retry_config.gas_increase_percent as u64;
+
+					warn!(
+						"Transaction failed (attempt {}/{}), retrying with {}% higher gas after {}ms",
+						attempt,
+						self.retry_config.max_attempts,
+						gas_buffer_adjustment,
+						delay
+					);
+
+					tokio::time::sleep(tokio::time::Duration::from_millis(delay)).await;
+				},
+			}
+		}
+	}
+
+	/// Calculate gas fees with additional buffer percentage
+	async fn calculate_gas_fees_with_buffer(
+		&self,
+		additional_buffer_percent: u64,
+	) -> Result<(U256, U256), ()> {
+		// Get current gas price from network
+		let current_gas_price = self
+			.rpc_client
+			.get_gas_price()
+			.await
+			.map_err(|_| error!("Failed to fetch gas price"))?;
+
+		// Apply base buffer plus additional retry buffer
+		let total_buffer = self.gas_config.gas_price_buffer_percent + additional_buffer_percent;
+		let buffer_multiplier = 100 + total_buffer;
+
+		let max_fee_per_gas = U256::from(current_gas_price)
+			.saturating_mul(U256::from(buffer_multiplier))
+			.checked_div(U256::from(100))
+			.unwrap_or(U256::from(current_gas_price));
+
+		// Calculate priority fee (tip) with bounds
+		let priority_fee = U256::from(current_gas_price / 10)
```

Comment:
Just wondering if that's the common way of estimation?

Reply:
Good question, this is the logic we currently have so I didn’t researched it. (I just added the `additional_buffer_percent` to the formula). To me it looks fine but I don't know if it is a common way of estimation.

I found that alloy provides this method: [estimate_eip1559_fees](https://docs.rs/alloy/latest/alloy/providers/trait.Provider.html#method.estimate_eip1559_fees), how about if we use it instead? c.c @kziemianek

Reply:
Done